File: Lua/counting_sort.lua
function counting_sort(arr)
    local max_val = math.max(table.unpack(arr))
    local min_val = math.min(table.unpack(arr))
    local range = max_val - min_val + 1

    local count = {}
    for i = 1, range do count[i] = 0 end

    for i = 1, #arr do
        count[arr[i] - min_val + 1] = count[arr[i] - min_val + 1] + 1
    end

    local index = 1
    for i = 1, range do
        while count[i] > 0 do
            arr[index] = i + min_val - 1
            count[i] = count[i] - 1
            index = index + 1
        end
    end
end

local arr = {5, 1, 4, 2, 8, 2, 1}
counting_sort(arr)
print(table.concat(arr, " "))

File: JavaScript/counting_sort.js
function countingSort(arr) {
    if (arr.length === 0) return arr;

    const max = Math.max(...arr);
    const min = Math.min(...arr);

    const count = new Array(max - min + 1).fill(0);

    for (let i = 0; i < arr.length; i++) {
        count[arr[i] - min]++;
    }

    let sortedIndex = 0;
    for (let i = 0; i < count.length; i++) {
        while (count[i] > 0) {
            arr[sortedIndex++] = i + min;
            count[i]--;
        }
    }

    return arr;
}

const arr = [4, 2, 2, 8, 3, 3, 1];
console.log(countingSort(arr));